### PR TITLE
Check roles are not empty

### LINKED ISSUE #15802: Roles with git submodule does nothing
##### ISSUE TYPE

- Feature Idea

##### COMPONENT NAME
"Playbook roles"

##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.0.1.0
  config file = ..../ansible.cfg
  configured module search path = Default w/o overrides
```
##### OS / ENVIRONMENT

N/A
##### SUMMARY

We have some git submodule in the roles repository, and sometimes we don't update our git clone submodules, which make some failures.
##### STEPS TO REPRODUCE

A teammate adds a new "role submodule"
git pull (creates an empty directory for the new submodule)
running a playbook that includes this new role does nothing, since roles is an empty directory

```
<!--- Paste example playbooks or commands between quotes -->
```

##### EXPECTED RESULTS

ansible-playbook should warn about that empty role, or maybe fail because a role that does nothing should not exists.
##### ACTUAL RESULTS

role is "ignored"

```
<!--- Paste verbatim command output between quotes -->
```

### PR DESCRIPTION
**Pull request by ovv**
##### SUMMARY

Verify a role is not empty. If it is error out.

It check if there is any tasks, handlers, metadata or others directories

Fixes #15802

##### ISSUE TYPE

  - Bugfix Pull Request

##### COMPONENT NAME

playbook/role/__init__.py
##### ANSIBLE VERSION

devel

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Review by robinro — APPROVED** [review]
LGTM

**Review by gundalow — COMMENTED** [review]
Unrelated changes in this PR

File: lib/ansible/playbook/role/__init__.py [hunk 1]
```diff
@@ -222,6 +222,10 @@ def _load_role_data(self, role_include, parent_role=None):
                 raise AnsibleParserError("The handlers/main.yml file for role '%s' must contain a list of tasks" % self._role_name,
                                          obj=handler_data, orig_exc=e)
 
+        directories = [directory for directory in self._loader.list_directory(self._role_path) if directory not in ('tasks', 'handlers', 'meta')]
```
Comment (bcoca) [inline lib/ansible/playbook/role/__init__.py:225]: roles are still valid as 'vars' providers, or files or templates or just a dependency grouping, etc .. not sure this is a good idea

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/playbook/role/__init__.py (hunk 1 above)
Comment (robinro) [reply lib/ansible/playbook/role/__init__.py:225]: For dependency-grouping there will be `metadata`, right?

The list of directories are those, that are already parsed by the role init: meta, tasks, handlers. All other directories are assumed to have non-trivial content and thus the role considered non-empty when they exist.

**Comment by jpmens** [comment]
Why does @ansibot cc half the Internet?

**Comment by willthames** [comment]
@jpmens, because this PR contains 85 files spread across various owners (I'm cc'd because of changes to some AWS tests, you're cc'd because of changes to the get_url test role) etc.

I'm sure you could work out the remaining cc list by mapping test changes to module owners.

**Comment by dagwieers** [comment]
@gundalow We shouldn't call it a "bad PR" because it changed more than 50 files. We have this protection to avoid spamming a lot of contributors (when it's having a lot of merge commits), not because the PR is intrinsically bad. We need to change this workflow.

**Comment by bcoca** [comment]
@dagwieers phrasing might need retouching, but this was created mostly due to PRs created from/pointing at the wrong branch and including many unrelated commits to many files as a result . That is why this check considers these 'bad PRs', but it is clearly not true for false positives.

**Comment by bcoca** [comment]
bot_skip

**Comment by dagwieers** [comment]
@bcoca There is more than one cause, in this case it was triggered because of the amount of files it covered, but there's another check for testing for merge commits. The second *could* be considered a "bad PR", but still the choice of words could be interpreted as a judgement of the code. I consider this something to put on the agenda for "Contributor Experience" hence I included @gundalow :-)

**Comment by bcoca** [comment]
@dagwieers yes, there is more than one reason you'll have 50 files, but that was the intention of the check, but it can still create false positives like this one.  Merge commits are clearly wrong so there are no false positives there. 

Since we CAN have false positives, I agree we should change the text on the check to reflect that.

**Comment by ovv** [comment]
I agree the wording could probably be better.
I decided to close this because I thought it was going nowhere and I don't really have the time to deal with it for the moment. If someone wants to take it over feel free to do so.

**Comment by bcoca** [comment]
@ovv I'm fine with closing it, I was mostly waiting for a response on my concerns about this change, since it also has many 'false postives' on what it considers invalid/empty roles

**Comment by dagwieers** [comment]
@bcoca Not trying to split hairs here, but all I wanted to point out is that whatever the cause (even if a PR includes merge commits), the PR should not be called "bad". Maybe call it "improperly constructed" and maybe provide feedback on how to recover your PR from those merge commits. (And that's what I meant with changing the workflow for the contributor) But I didn't want to have the discussion in this "good" PR :-D

**Comment by bcoca** [comment]
@dagwieers i think we don't have any disagreement, sorry it seemed that way, I was only trying to clarify the target of the checks.

**Comment by gundalow** [comment]
Hi,
Please raise issues under `ansible/ansibullbot/` and link this PR. Discussing process in a PR means the important work of the PR gets lost in the back-and-forth discussions of PR work flow/bot bugs..

@jpmens New bot functionality will not add labels or `@mention` people if there are more than 50 files changes, so this should reduce the notification spam.

File: lib/ansible/playbook/role/__init__.py (hunk 1 above)
Comment (bcoca) [reply lib/ansible/playbook/role/__init__.py:225]: roles can just provide 'content' , vars/defaults/files/templates for use in other things, so to be really empty it must have NOTHING, this only checks for a subset of that.

**Comment by gundalow** [comment]
Hi,
Thank you for taking the time to improve Ansible. 

Given the discussion above that we'd need to check the role is "truly empty" which would be a lot more work I'm going to close this.

**Pull request closed without merging**